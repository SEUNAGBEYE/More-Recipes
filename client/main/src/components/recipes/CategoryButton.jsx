import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';


/**
 * @description - Component for selecting category to display
 * @param {Object} props
 *
 * @returns {jsx} JSX
 */
const CategoryButton = (props) => (
  <div className="dropdown" style={{ float: 'left' }}
    data-category-button="category-button"
  >
    <button className="btn btn-default dropdown-toggle auth-button"
      type="button" id="about-us" data-toggle="dropdown"
      aria-haspopup="true" aria-expanded="false">Category
    </button>
    <div className="dropdown-menu" aria-labelledby="about-us">
      {
        props.categories.map(category =>
          (<Link className="dropdown-item"
            key={category.id} to={`/categories/${category.name}`}
            data-category-link={`category-link-${category.name}`}
          >{category.name}</Link>))
      }
    </div>
  </div>
);

const propTypes = {
  categories: PropTypes.array.isRequired,
};

CategoryButton.propTypes = propTypes;

/**
 * mapStateToProps
 * @export
 * @param {Object} state
 *
 * @returns {Object} object
 */
export const mapStateToProps = (state) => ({
  categories: state.recipes.recipeCategories
    .map(category => ({ name: category.name, id: category.id }))
});

export { CategoryButton as CategoriesButton };
export default connect(mapStateToProps)(CategoryButton);
