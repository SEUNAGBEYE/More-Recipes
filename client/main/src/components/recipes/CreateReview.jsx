import React from 'react';
import PropTypes from 'prop-types';

/**
 * @description - Component displays form for creating a review
 * @param {Object} props
 *
 * @returns {jsx} JSX
 */
const CreateReview = (props) => (
  <form className="form-group" id="create-review-form"
    onSubmit={props.reviewRecipe}
  >

    <textarea rows="5" cols="20" className="form-control"
      placeholder="Write reviews here" name="reviewBody"
      id="review"
      value={props.reviewBody}
      onChange={props.onChange}
    />
    <button className="btn-default auth-button"
      style={{ margin: 10, float: 'left' }}
    >
     Submit
    </button>
  </form>
);

const propTypes = {
  reviewRecipe: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  reviewBody: PropTypes.string.isRequired,
};

CreateReview.propTypes = propTypes;

export default CreateReview;

