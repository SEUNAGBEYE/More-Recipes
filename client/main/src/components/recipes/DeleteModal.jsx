import React from 'react';
import PropTypes from 'prop-types';

const DeleteModal = (props) => (
  <div>
    <div className="modal fade" id={`deleteModal${props.id}`}
      tabIndex="-1" role="dialog" aria-labelledby="modalLabel"
      aria-hidden="true"
    >
      <div className="modal-dialog" role="document">
        <div className="modal-content">
          <div className="modal-header">
            <button type="button" className="close" data-dismiss="modal"
              aria-label="Close" style={{ color: 'orange' }}
            >
              <span aria-hidden="true">&times;</span>
            </button>
            <h4 className="modal-title" id="modalLabel"
              style={{ color: 'orange' }}
            >
            Delete
            </h4>
          </div>

          <div className="modal-body">
            <p style={{ fontSize: 16 }}>
              <strong>Do you want to delete this recipe?</strong>
            </p>
          </div>

          <div className="modal-footer">
            <button className="btn btn-secondary auth-button"
              data-dismiss = "modal" onClick={props.onDelete}
              id={props.id}>Yes
            </button>
            <button type="button" className="btn btn-secondary auth-button"
              data-dismiss="modal"
            >
                  No
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
);

const propTypes = {
  id: PropTypes.number.isRequired,
  onDelete: PropTypes.func.isRequired,
};

DeleteModal.propTypes = propTypes;

export default DeleteModal;
