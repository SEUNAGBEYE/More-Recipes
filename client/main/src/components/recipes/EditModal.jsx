import React, { Component } from 'react';
import uuid from 'uuid';
import Loader from 'react-loader';
import PropTypes from 'prop-types';
import Input from './Input';
import setAuthorizationToken from '../../../utils/setAuthorizationToken';
import imageUpload from '../../../utils/ImageUploader';
/**
 * @export
 * @class EditModal
 * @extends {Component}
 */
export default class EditModal extends Component {
  /**
  * Creates an instance of EditModal.
  * @param {obj} props
  * @memberof EditModal
  */
  constructor(props) {
    super(props);
    const { recipe } = this.props;
    const {
      steps, ingredients, name, description, image
    } = recipe;
    this.state = {
      name,
      description,
      image,
      ingredients,
      steps,
      errors: {},
      stepsTimes: '',
      ingredientsTimes: [],
      loaded: true,
    };
    this.onChange = this.onChange.bind(this);
    this.stepClick = this.stepClick.bind(this);
    this.ingredientClick = this.ingredientClick.bind(this);
    this.updateRecipe = this.updateRecipe.bind(this);
    this.removeIngredientsInput = this.removeIngredientsInput.bind(this);
    this.removeStepsInput = this.removeStepsInput.bind(this);
  }

  /**
   * @param {Object} event
   *
   * @returns {void} void
   * * @memberof EditModal
   */
  stepClick(event) {
    event.preventDefault();
    this.setState({ steps: [...this.state.steps, ''] });
  }

  /**
   * @param {Object} event
   *
   *  @returns {void} void
   * @memberof EditModal
   */
  ingredientClick(event) {
    event.preventDefault();
    this.setState({ ingredients: [...this.state.ingredients, ''] });
  }

  /**
   *@description - Remove form to add ingredients
   * @param {any} event
   *
   * @returns {void} void
   * @memberof EditModal
   */
  removeIngredientsInput(event) {
    const { ingredients } = this.state;
    const { index: eventIndex } = event.target.dataset;
    const newIngredients = ingredients
      .filter((ingredient, index) => parseInt(eventIndex, 10) !== parseInt(index, 10));
    this.setState({ ingredients: newIngredients });
  }

  /**
   * @description - Remove form to add a step
   * @param {Object} event
   *
   * @returns {void} void
   * @memberof EditModal
   */
  removeStepsInput(event) {
    const { steps } = this.state;
    const { index: eventIndex } = event.target.dataset;
    const newSteps = steps
      .filter((step, index) => parseInt(eventIndex, 10) !== parseInt(index, 10));
    this.setState({ steps: newSteps });
  }


  /**
   * @param {Object} event
   *
   *  @returns {void} void
   * @memberof EditModal
   */
  onChange(event) {
    event.preventDefault();
    const stateKeyIndex = event.target.dataset.index;
    const { name: stateKey, value } = event.target;
    if (stateKey === 'steps' || stateKey === 'ingredients') {
      this.setState({
        [event.target.name]: this.state[stateKey].map((step, index) => {
          if (parseInt(index, 10) === parseInt(stateKeyIndex, 10)) {
            step = value;
          }
          return step;
        })
      });
    } else {
      this.setState({
        [stateKey]: value
      });
    }
  }

  /**
   * @param {Object} event
   *
   *  @returns {void} void
   * @memberof EditModal
   */
  updateRecipe(event) {
    event.preventDefault();
    const { id } = event.target;
    const file = document.getElementById(`recipePicture${id}`).files[0];
    if (file) {
      if (file.size > 4000000) {
        return toastr.warning('File too Large');
      } else {
        this.setState({ loaded: false });
        imageUpload(file)
          .then(res => {
            this.setState({ image: res.data.secure_url }, () => {
              setAuthorizationToken(localStorage.token);
              this.props.editRecipe(this.props.recipe.id, this.state);
              this.setState({ loaded: true });
              $('.modal').modal('hide');
            });
          })
          .catch(error => {
            this.setState({ loaded: true });
          });
      }
    } else {
      this.setState({ loaded: true, image: this.props.recipe.image });
      this.props.editRecipe(this.props.recipe.id, this.state);
      $('.modal').modal('hide');
    }
  }

  /**
   * @returns {jsx} jsx
   * @memberof EditModal
   */
  render() {
    const { recipe } = this.props;
    const stepFields = this.state.steps.map((step, index) => (
      <Input key={uuid()}
        onChange={this.onChange}
        number={index + 1}
        id={uuid()}
        index={index}
        value={step}
        name={'steps'}
        removeInput={this.removeStepsInput}
      />
    ));

    const ingredientFields = this.state.ingredients.map((ingredient, index) => (
      <Input key={uuid()}
        onChange={this.onChange}
        number={index + 1}
        id={uuid()}
        index={index}
        value={ingredient}
        name={'ingredients'}
        removeInput={this.removeIngredientsInput}
      />
    ));

    return (
      <div>
        <div className="modal fade" id={`editModal${recipe.id}`}
          tabIndex="-1" role="dialog" aria-labelledby="modalLabel"
          aria-hidden="true"
        >
          <div className="modal-dialog" role="document">
            <div className="modal-content">
              <div className="modal-header">
                <button type="button" className="close" data-dismiss="modal"
                  aria-label="Close"
                >
                  <span aria-hidden="true">&times;</span>
                </button>
                <h4 className="modal-title" id="modalLabel">Recipe</h4>
              </div>

              <div className="modal-body">
                <form>
                  <fieldset className="form-group">
                    <label htmlFor="name" className="form-inline">Name</label>
                    <input type="text" className="form-control" id="recipeName"
                      name="name" onChange={this.onChange}
                      defaultValue={recipe.name}
                    />
                  </fieldset>

                  <fieldset className="form-group">
                    <label htmlFor="recipeDescription"
                      className="form-inline"
                    >
                      Description
                    </label>
                    <textarea className="form-control" id="description"
                      name="description" cols="50" rows = "5"
                      onChange={this.onChange}
                      defaultValue={recipe.description}
                    />
                  </fieldset>

                  <fieldset className="form-group">
                    <label htmlFor="image" className="form-inline">
                      Picture
                    </label>
                      The maximum file size allowed is 4mb
                    <input type="file" className="form-control"
                      id={`recipePicture${recipe.id}`} name="image"
                    />
                  </fieldset>

                  {ingredientFields}
                  <fieldset>
                    <button className="btn btn-default show-input fa fa-plus"id="ingredient" onClick={this.ingredientClick}><strong>Ingredients</strong></button>
                  </fieldset>

                  {stepFields}
                  <fieldset>
                    <button className="btn btn-default show-input fa fa-plus"id="step" onClick={this.stepClick}><strong>Steps</strong></button>
                  </fieldset>

                  <div className="modal-footer">
                    <Loader loaded={this.state.loaded} />
                    <button className="btn btn-secondary auth-button"
                      id={recipe.id} onClick={this.updateRecipe}
                      data-update={`update-recipe${recipe.id}`}
                    >
                    Update
                    </button>
                    <button type="button" className="btn btn-secondary auth-button" data-dismiss="modal">
                      Cancel
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>

    );
  }
}

const propTypes = {
  editRecipe: PropTypes.func.isRequired,
  recipe: PropTypes.object.isRequired,
};

EditModal.propTypes = propTypes;

