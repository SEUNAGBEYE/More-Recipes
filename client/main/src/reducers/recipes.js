import actionTypes from '../actions/actionTypes';

const {
  GET_RECIPE,
  GET_RECIPES,
  GET_USER_RECIPES,
  GET_USER_FAVOURITED_RECIPES_ID,
  GET_FAVOURITED_RECIPES,
  GET_POPULAR_RECIPES,
  GET_RECIPE_REVIEWS,
  RECIPE_CATEGORIES,
  SEARCH_RECIPES,
  ADD_RECIPE,
  DELETE_RECIPE,
  EDIT_RECIPE,
  TOGGLE_FAVOURITE_RECIPE,
  TOGGLE_THUMBS_UP_RECIPE,
  TOGGLE_THUMBS_DOWN_RECIPE,
  REVIEW_RECIPE
} = actionTypes;

export const initialState = {
  allRecipes: [],
  singleRecipe: {
    id: '',
    name: '',
    image: '',
    categoryId: '',
    userId: '',
    upvotes: [],
    downvotes: [],
    views: [],
    description: '',
    reviews: []
  },
  pagination: '',
  userFavouritedRecipeId: [],
  favouriteRecipes: [],
  popularRecipes: [],
  recipeCategories: [],
  loaded: false
};

export default (state = initialState, action = {}) => {
  const singleRecipe = { ...state.singleRecipe };
  let allRecipes = [...state.allRecipes];
  const { recipe: recipeOnAction } = action;

  switch (action.type) {
  case ADD_RECIPE:
    return {
      ...state,
      ...{ allRecipes: [action.recipe, ...state.allRecipes] }
    };

  case GET_RECIPE:
    return {
      ...state,
      ...{
        singleRecipe: action.recipe,
        pagination: action.pagination,
        loaded: true
      }
    };

  case GET_USER_RECIPES:
    return {
      ...state,
      ...{
        allRecipes: action.recipes,
        pagination: action.pagination
      }
    };

  case GET_RECIPES:
    return {
      ...state,
      ...{
        allRecipes: action.recipes,
        pagination: action.pagination,
        loaded: true
      }
    };

  case RECIPE_CATEGORIES:
    return {
      ...state,
      ...{
        recipeCategories: action.recipeCategories,
        loaded: true
      }
    };
  case GET_POPULAR_RECIPES:
    return {
      ...state,
      ...{
        popularRecipes: [...action.popularRecipe],
        loaded: true
      }
    };

  case SEARCH_RECIPES:
    return {
      ...state,
      ...{
        allRecipes: [...action.recipes],
        pagination: action.pagination
      }
    };

  case DELETE_RECIPE:
    return {
      ...state,
      ...{
        allRecipes: state.allRecipes
          .filter(recipe => parseInt(recipe.id, 10) !== parseInt(action.id, 10))
      }
    };


  case EDIT_RECIPE:
    if (allRecipes.length >= 1) {
      allRecipes = allRecipes.map((recipe) => {
        if (parseInt(recipe.id, 10) === parseInt(action.recipe.id, 10)) {
          return action.recipe;
        }
        return recipe;
      });
    }

    return {
      ...state,
      ...{
        allRecipes,
        singleRecipe: { ...singleRecipe, ...recipeOnAction }
      }
    };

  case TOGGLE_FAVOURITE_RECIPE:
    return {
      ...state,
      ...{
        favouriteRecipes: !state.userFavouritedRecipeId
          .includes(action.favouritedRecipe.id) ?
          [...state.favouriteRecipes, action.favouritedRecipe] :
          state.favouriteRecipes
            .filter(recipe => recipe.id != action.favouritedRecipe.id),

        userFavouritedRecipeId: !state.userFavouritedRecipeId
          .includes(action.favouritedRecipe.id) ?
          state.userFavouritedRecipeId.concat(action.favouritedRecipe.id) :
          state.userFavouritedRecipeId
            .filter(id => id !== action.favouritedRecipe.id),
        singleRecipe: action.favouritedRecipe
      }
    };

  case TOGGLE_THUMBS_UP_RECIPE:
    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        singleRecipe: action.recipe
      }
    };

  case TOGGLE_THUMBS_DOWN_RECIPE:
    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            return action.recipe;
          }
          return recipe;
        })],
        singleRecipe: action.recipe
      }
    };

  case GET_FAVOURITED_RECIPES:
    return {
      ...state,
      favouriteRecipes: [...action.favouriteRecipes],
      pagination: action.pagination
    };

  case GET_USER_FAVOURITED_RECIPES_ID:
    return {
      ...state,
      ...{ userFavouritedRecipeId: action.favouritedRecipesIds }
    };
  case REVIEW_RECIPE:
    singleRecipe.reviews = [action.reviewBody, ...singleRecipe.reviews];
    return {
      ...state,
      ...{ singleRecipe }
    };
  case GET_RECIPE_REVIEWS:
    singleRecipe.reviews = [...singleRecipe.reviews, ...action.reviews];
    return {
      ...state,
      ...{
        singleRecipe,
        pagination: action.pagination
      }
    };

  default: return {
    ...state,
    ...{ loaded: true }
  };
  }
};

export { actionTypes };
