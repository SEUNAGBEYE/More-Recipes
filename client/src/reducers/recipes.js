
import isEmpty from 'lodash/isEmpty';

const initialState = {
  allRecipes: [],
  recipesCount: '',
  userFavouritedRecipeId: [],
  favouriteRecipes: [],
  popularRecipes: []
  // favouritedRecipesCount: '',
  // userRecipes: []
};

export default (state = initialState, action = {}) => {
  switch (action.type) {
  case 'ADD_RECIPE':
    return [...state, action.recipe];

  case 'GET_RECIPE':
    return { ...state, ...{ allRecipes: [...state.allRecipes.filter(recipe => recipe.id === parseInt(action.id))] || action.recipe } };

  case 'GET_USER_RECIPES':
    return {
      ...state,
      ...{
        allRecipes:
        action.recipes
      }
    };

  case 'GET_RECIPES':

    return {
      ...state,
      ...{
        allRecipes: action.allRecipes,
        recipesCount: action.recipesCount
      }
    };
  case 'GET_POPULAR_RECIPES':
    console.log('popular', action.popularRecipe);
    return {
      ...state,
      ...{
        popularRecipes: [...action.popularRecipe]
      }
    };

  case 'DELETE_RECIPE':
    return state.allRecipe.filter(recipe => recipe.id !== action.id);


  case 'EDIT_RECIPE':

    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id == action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })]
      }
    };

  case 'TOGGLE_FAVOURITE_RECIPE':
    return {
      ...state,
      ...{
        favouriteRecipes: !state.userFavouritedRecipeId.includes(action.favouritedRecipe.id)
          ? [...state.favouriteRecipes, action.favouritedRecipe]
          : state.favouriteRecipes.filter(recipe => recipe.id != action.favouritedRecipe.id),

        userFavouritedRecipeId: !state.userFavouritedRecipeId.includes(action.favouritedRecipe.id)
          ? state.userFavouritedRecipeId.concat(action.favouritedRecipe.id)
          : state.userFavouritedRecipeId.filter(id => id !== action.favouritedRecipe.id)
      }
    };

  case 'TOGGLE_THUMBS_UP_RECIPE':


    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
      }
    };

  case 'TOGGLE_THUMBS_DOWN_RECIPE':

    return {
      ...state,
      ...{
        allRecipes: [...state.allRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
        favouriteRecipes: [...state.favouriteRecipes.map((recipe) => {
          if (recipe.id === action.recipe.id) {
            recipe = action.recipe;
          }
          return recipe;
        })],
      }
    };

  case 'GET_FAVOURITED_RECIPES':
    return {
      ...state,
      favouriteRecipes: [...action.favouriteRecipes],
      recipesCount: action.favouritedRecipesCount
    };

  case 'GET_USER_FAVOURITED_RECIPES_ID':
    return {
      ...state,
      ...{ userFavouritedRecipeId: action.favouritedRecipesIds }
    };

  default: return state;
  }
};
