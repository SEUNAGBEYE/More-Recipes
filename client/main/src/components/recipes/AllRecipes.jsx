import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  getFavouritedRecipesIds,
  addRecipe,
  allRecipes,
  recipeCategories
}
  from '../../actions/Recipes';
import RecipeCard from './RecipeCard';
import Pagination from './Pagination';
import Exclamation from './Exclamation';
import RecipeModal from './RecipeModal';
import CategoryButton from './CategoryButton';

/**
 * @export
 * @class AllRecipes
 * @extends {Component}
 */
export class AllRecipes extends Component {
  /**
   * Creates an instance of AllRecipes.
   * @param {Object} props
   * @memberof AllRecipes
   */
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.paginateRecipes = this.paginateRecipes.bind(this);
  }

  /**
   * @returns {void} void
 * @memberof AllRecipes
 */
  componentDidMount() {
    this.props.recipeCategories();
    this.paginateRecipes(1);
  }

  /**
   * @param {Number} page
   *
   * @returns {void} void
   * @memberof AllRecipes
   */
  paginateRecipes(page) {
    this.props.allRecipes(page);
  }


  /**
   * @param {Object} data
   *
   * @returns {void} void
   * @memberof AllRecipes
   */
  onSubmit(data) {
    this.props.addRecipe(data)
      .then(res => {
        this.setState({ allRecipes: [...this.state.allRecipes, res.recipe] });
      });
  }

  /**
   * @memberOf UserRecipes
   * @returns {jsx} JSX
   * return {object}
   */
  render() {
    return (
      <div>
        <main style={{ marginTop: 40 }}>

          <div className="container">
            <CategoryButton/>
          </div>

          <div className="container">
            <div style={{ textAlign: 'center', marginTop: 100 }}>
              <h4 className="container__myrecipes">All Recipes</h4><br /><br />
              <RecipeModal
                addRecipe={this.onSubmit}
                recipeCategories={this.props.categories}
              />
            </div>

            <div className="row">
              {
                this.props.recipes.length > 0 ?
                  this.props.recipes.map((recipe, index) => (
                    <RecipeCard key={recipe.id}
                      recipe={recipe}
                      id={recipe.id}
                      history={this.props.history}
                    />)) :
                  <Exclamation>
                    <p className="text-muted text-center">
                      Sorry no recipe has been added yet,
                      please add to get started
                    </p>
                  </Exclamation>
              }
            </div>
          </div>
        </main>
        {this.props.pagination > 1 ? <Pagination recipesCount={this.props.pagination}
          recipesPagination={this.paginateRecipes}/> :
          ''
        }
      </div>
    );
  }
}

const propTypes = {
  recipes: PropTypes.array.isRequired,
  addRecipe: PropTypes.func.isRequired,
  recipeCategories: PropTypes.func.isRequired,
  allRecipes: PropTypes.func.isRequired,
  categories: PropTypes.array.isRequired,
  pagination: PropTypes.number.isRequired,
  history: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired
};

AllRecipes.propTypes = propTypes;

/**
 * mapStateToProps
 * @export
 * @param {Object} state
 *
 * @returns {Object} object
 */
export const mapStateToProps = (state) => ({
  recipes: state.recipes.allRecipes,
  pagination: Number(state.recipes.pagination),
  user: state.auth.user,
  isAuthenticated: state.auth.isAuthenticated,
  categories: state.recipes.recipeCategories
});

export default connect(mapStateToProps, {
  getFavouritedRecipesIds,
  allRecipes,
  addRecipe,
  recipeCategories
})(AllRecipes);

