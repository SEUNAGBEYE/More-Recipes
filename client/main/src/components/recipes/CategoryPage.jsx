import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Loader from 'react-loader';
import { recipeCategories } from '../../actions/Recipes';
import RecipeCard from './RecipeCard';
import Pagination from './Pagination';
import Exclamation from './Exclamation';
import CategoryButton from './CategoryButton';


/**
 * @class CategoryPage
 * @extends {Component}
 */
export class CategoryPage extends Component {
  /**
   * @description - Creates an instance of AllRecipes.
   *
   * @method contructor
   *
   * @param {any} props
   *
   * @returns {void} void
   * @memberof AllRecipes
   */
  constructor(props) {
    super(props);
    this.paginateRecipes = this.paginateRecipes.bind(this);
    this.state = {
      loaded: false
    };
  }

  /**
   * @description - Get component data
   *
   * @method componentDidMount
   *
   * @returns {void} void
   * @memberof CategoryPage
   */
  componentDidMount() {
    this.paginateRecipes(1);
  }

  /**
   * @description - Get all recipes
   *
   * @method paginateRecipe
   *
   * @param {Number} page
   *
   * @returns {void} void
   * @memberof CategoryPage
   */
  async paginateRecipes(page) {
    await this.props.recipeCategories();
    this.setState({ loaded: true });
  }

  /**
   * @description - Renders react component
   *
   * @returns {Jsx} Jsx
   * @memberof CategoryPage
   */
  render() {
    return (
      <div>
        { this.state.loaded ?
          <main style={{ marginTop: 40 }}>
            <div className="container">
              <CategoryButton />
            </div>

            <div className="container">
              <div style={{ textAlign: 'center', marginTop: 100 }}>
                <h4 className="container__myrecipes">
                  {this.props.category.name}
                </h4><br /><br />
              </div>

              <div className="row">
                {
                  this.props.category.recipes.length > 0 ?
                    this.props.category.recipes.map((elem, index) => (
                      <RecipeCard key={elem.id}
                        recipe={elem}
                        id={elem.id}
                        history={this.props.history}
                      />)) :
                    <Exclamation>
                      <p className="text-muted text-center">
                        {
                          `Sorry no recipe has been added to 
                        ${this.props.category.name} 
                       yet, please add to get started`
                        }
                      </p>
                    </Exclamation>
                }
              </div>
            </div>
            { this.props.pagination > 1 ?
              <Pagination recipesCount={this.props.pagination}
                recipesPagination={this.paginateRecipes}
              /> : ''
            }
          </main> :
          <Loader loaded={this.state.loaded}/> }
      </div>
    );
  }
}

/**
 * mapStateToProps
 * @param {Object} state
 * @param {Object} props
 *
 * @returns {Object} object
 */
export const mapStateToProps = (state, props) => ({
  recipes: state.recipes.allRecipes,
  pagination: Number(state.recipes.pagination),
  isAuthenticated: state.auth.isAuthenticated,
  category: state.recipes.recipeCategories
    .find(recipeCategory => (
      recipeCategory.name === props.match.params.categoryName
    )) || {}
});

const propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  recipes: PropTypes.array.isRequired,
  category: PropTypes.object.isRequired,
  recipeCategories: PropTypes.func.isRequired,
  pagination: PropTypes.number.isRequired,
  history: PropTypes.object.isRequired

};

CategoryPage.propTypes = propTypes;

export default connect(mapStateToProps, { recipeCategories })(CategoryPage);

