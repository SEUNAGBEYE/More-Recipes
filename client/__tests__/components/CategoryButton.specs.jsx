
import React from 'react';
import { shallow } from 'enzyme';
import { CategoriesButton as CategoryButton, mapStateToProps } from '../../main/src/components/recipes/CategoryButton';

const state = {
  recipes: {
    recipeCategories: [{
      id: 1,
      name: 'Breakfast'
    }]
  }
};

const recipe = {
  name: 'Amala',
  image: 'This is the image'
}

const props = {
  categories: [],
};

describe('# CategoryButton', () => {
  const wrapper = shallow(
      <CategoryButton {...props} />
  );
  it('should render successfully', (done) => {
    expect(wrapper).toBeDefined();
    expect(wrapper.getElement().type).toBe('div');
    expect(wrapper).toMatchSnapshot()
    done();
  });

  it('should call mapStateToProps(state)', (done) => {
    mapStateToProps(state)
    done();
  }); 
});
