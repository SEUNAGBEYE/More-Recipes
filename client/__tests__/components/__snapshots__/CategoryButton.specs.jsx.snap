// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# CategoryButton should render successfully 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CategoryButton
    categories={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          aria-expanded="false"
          aria-haspopup="true"
          className="btn btn-default dropdown-toggle auth-button"
          data-toggle="dropdown"
          id="about-us"
          type="button"
        >
          Category
        </button>,
        <div
          aria-labelledby="about-us"
          className="dropdown-menu"
        />,
      ],
      "className": "dropdown",
      "data-category-button": "category-button",
      "style": Object {
        "float": "left",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-expanded": "false",
          "aria-haspopup": "true",
          "children": "Category",
          "className": "btn btn-default dropdown-toggle auth-button",
          "data-toggle": "dropdown",
          "id": "about-us",
          "type": "button",
        },
        "ref": null,
        "rendered": "Category",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-labelledby": "about-us",
          "children": Array [],
          "className": "dropdown-menu",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            aria-expanded="false"
            aria-haspopup="true"
            className="btn btn-default dropdown-toggle auth-button"
            data-toggle="dropdown"
            id="about-us"
            type="button"
          >
            Category
          </button>,
          <div
            aria-labelledby="about-us"
            className="dropdown-menu"
          />,
        ],
        "className": "dropdown",
        "data-category-button": "category-button",
        "style": Object {
          "float": "left",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-expanded": "false",
            "aria-haspopup": "true",
            "children": "Category",
            "className": "btn btn-default dropdown-toggle auth-button",
            "data-toggle": "dropdown",
            "id": "about-us",
            "type": "button",
          },
          "ref": null,
          "rendered": "Category",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-labelledby": "about-us",
            "children": Array [],
            "className": "dropdown-menu",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
